src/main/scala/linting/rule/RenameDesiredNames.scala

```diff
@@ -0,0 +1,197 @@
+
+/** A helper to rename modules in a [[Circuit]]
+  */
+object RenameModules {
+  def onStmt(moduleNameMap: Map[String, String])(stmt: Statement): Statement = stmt match {
+    case inst: WDefInstance if moduleNameMap.contains(inst.module) => inst.copy(module = moduleNameMap(inst.module))
+    case inst: DefInstance if moduleNameMap.contains(inst.module) => inst.copy(module = moduleNameMap(inst.module))
+    case other => other.mapStmt(onStmt(moduleNameMap))
+  }
+
+  /** Renames the modules in a circuit given a mapping of old names to new names
+    *
+    * @param nameMappings mapping of old to new names
+    * @param circuit the circuit to rename
+    */
+  def apply(nameMappings: Map[String, String], circuit: Circuit): Circuit = {
+    val modules = circuit.modules.map {
+      case mod: Module => mod.mapStmt(onStmt(nameMappings)).mapString(m => nameMappings.getOrElse(m, m))
+      case ext: ExtModule => ext
+    }
+    val main = nameMappings.getOrElse(circuit.main, circuit.main)
+    circuit.copy(main = main, modules = modules)
+  }
+}
+
+/** A strategy for generating a new names for modules
+  */
+sealed trait NamingStrategy {
+
+  /** Generates a new stable module name based on the desired name and the module IR node
+    *
+    * called by [[LintConflictingModuleNames]] to rename modules
+    *
+    * @param desiredName the requested name for the module, generated name should contain this name
+    * @param module the module targeted by the desiredName
+    */
+  def getName(desiredName: String)(module: Module): String
+}
+
+/** A naming strategy that always returns the desired name
+  */
+case object ExactNamingStrategy extends NamingStrategy {
+  def getName(desiredName: String)(module: Module): String = desiredName
+}
+
+/** Specifies a naming strategy to use for a module and the modules that it collides with
+  */
+case class NamingStrategyAnnotation(
+  strategy: NamingStrategy,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] with HasSerializationHints {
+  def typeHints: Seq[Class[_]] = Seq(
+    ExactNamingStrategy.getClass
+  )
+
+  def duplicate(newTarget: IsModule): NamingStrategyAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Specifies the desired name to rename the module to, overriding the Module.desiredName
+  */
+case class OverrideDesiredNameAnnotation(
+  desiredName: String,
+  target: IsModule
+) extends SingleTargetAnnotation[IsModule] {
+  def duplicate(newTarget: IsModule): OverrideDesiredNameAnnotation = {
+    this.copy(target = newTarget)
+  }
+}
+
+/** Renames modules based on their overridden desired names
+  *
+  * Desired module name overrides are specified by
+  * [[OverrideDesiredNameAnnotation]]. If one of the modules is targeted by a
+  * [[NamingStrategyAnnotation]] then the transform will attempt to rename
+  * modules according to that naming strategy. The default naming strategy is
+  * [[ExactNamingStrategy]] which will throw an exception if there are any
```

Comment:
Sorry I think I already asked this before, but wouldn't linter take care of this?

Reply:
or is the point that for ExactNamingStrategy,  you get an exception, wehreas if you came up with a better strategy you wouldn't get an exception, you would just hash to get a unique name. That makes sense.

Reply:
sorry yes linter takes care of this. scaladoc is out of date